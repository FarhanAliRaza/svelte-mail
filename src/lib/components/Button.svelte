<script lang="ts">
	import type { Snippet } from 'svelte';
	import type { ClassValue, HTMLAnchorAttributes } from 'svelte/elements';

	/**
	 * Computes optimal font width and space count for MSO padding simulation
	 */
	function computeFontWidthAndSpaceCount(expectedWidth: number) {
		if (expectedWidth === 0) return [0, 0];

		let smallestSpaceCount = 0;
		const maxFontWidth = 5;

		const computeRequiredFontWidth = () => {
			if (smallestSpaceCount > 0) {
				return expectedWidth / smallestSpaceCount / 2;
			}
			return Number.POSITIVE_INFINITY;
		};

		while (computeRequiredFontWidth() > maxFontWidth) {
			smallestSpaceCount++;
		}

		return [computeRequiredFontWidth(), smallestSpaceCount];
	}

	type PaddingType = string | number | undefined;
	function convertToPx(value: PaddingType) {
		if (!value) return 0;
		if (typeof value === 'number') return value;
		const matches = /^([\d.]+)(px|em|rem|%)$/.exec(value);
		if (matches && matches.length === 3) {
			const numValue = Number.parseFloat(matches[1]);
			switch (matches[2]) {
				case 'px':
					return numValue;
				case 'em':
				case 'rem':
					return numValue * 16;
				case '%':
					return (numValue / 100) * 600;
				default:
					return numValue;
			}
		}
		return 0;
	}
	function parsePadding(props: any) {
		let pt = 0,
			pr = 0,
			pb = 0,
			pl = 0;
		let { padding = '', paddingTop, paddingRight, paddingBottom, paddingLeft } = props || {};
		if (typeof padding === 'number') {
			pt = pr = pb = pl = padding;
		} else if (typeof padding === 'string') {
			const values = padding.split(/\s+/);
			switch (values.length) {
				case 1:
					pt = pr = pb = pl = convertToPx(values[0]);
					break;
				case 2:
					pt = pb = convertToPx(values[0]);
					pr = pl = convertToPx(values[1]);
					break;
				case 3:
					pt = convertToPx(values[0]);
					pr = pl = convertToPx(values[1]);
					pb = convertToPx(values[2]);
					break;
				case 4:
					pt = convertToPx(values[0]);
					pr = convertToPx(values[1]);
					pb = convertToPx(values[2]);
					pl = convertToPx(values[3]);
					break;
			}
		}
		return {
			pt: paddingTop ? convertToPx(paddingTop) : pt,
			pr: paddingRight ? convertToPx(paddingRight) : pr,
			pb: paddingBottom ? convertToPx(paddingBottom) : pb,
			pl: paddingLeft ? convertToPx(paddingLeft) : pl
		};
	}

	// Parse padding from style string if present
	function parsePaddingFromStyle(style: string | undefined) {
		if (!style || typeof style !== 'string') return {};
		const result: Record<string, string> = {};
		const regex = /padding(-top|-right|-bottom|-left)?:\s*([^;]+);?/gi;
		let match;
		while ((match = regex.exec(style))) {
			const key = match[1]
				? 'padding' + match[1].replace(/-([a-z])/, (_, c) => c.toUpperCase())
				: 'padding';
			result[key] = match[2].trim();
		}
		return result;
	}

	interface Props extends HTMLAnchorAttributes {
		class?: ClassValue;
		children: Snippet;
		padding?: string | number;
		paddingLeft?: string | number;
		paddingRight?: string | number;
		paddingTop?: string | number;
		paddingBottom?: string | number;
		[key: string]: any;
	}

	let {
		href = '',
		target = '_blank',
		style,
		class: className,
		children,
		padding = 0,
		paddingLeft,
		paddingRight,
		paddingTop,
		paddingBottom,
		...restProps
	}: Props = $props();

	// Parse padding from style string if present
	const stylePadding = parsePaddingFromStyle(style as string | undefined);

	// Use explicit props if provided, otherwise fallback to style-derived values
	const paddingObj = parsePadding({
		padding: padding ?? stylePadding.padding,
		paddingLeft: paddingLeft ?? stylePadding.paddingLeft,
		paddingRight: paddingRight ?? stylePadding.paddingRight,
		paddingTop: paddingTop ?? stylePadding.paddingTop,
		paddingBottom: paddingBottom ?? stylePadding.paddingBottom
	});

	// Individual padding values override the shorthand if provided
	const pt = paddingObj.pt;
	const pr = paddingObj.pr;
	const pb = paddingObj.pb;
	const pl = paddingObj.pl;

	// Calculate vertical spacing for MSO text raise
	const y = pt + pb;
	const pxToPt = (px: number) => Math.floor(px * 0.75);

	const textRaise = pxToPt(y);
	// Calculate font width and space count for MSO padding simulation
	const [plFontWidth, plSpaceCount] = computeFontWidthAndSpaceCount(pl);
	const [prFontWidth, prSpaceCount] = computeFontWidthAndSpaceCount(pr);

	// Build the final style object

	const buttonStyle =
		'line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; padding: 0px 0px 0px 0px; mso-padding-alt: 0px;';

	// Ensure target="_blank" links are secure
	const rel = target === '_blank' ? 'noopener noreferrer' : undefined;

	// For MSO hack, use unicode 8202 (hair space) for spacing
	const zeroWidthSpace = '&#8202;';
</script>

<a {href} {target} {rel} style="{buttonStyle} {style}" class={className} {...restProps}>
	<span>
		{@html `<!--[if mso]><i style="mso-font-width:${plFontWidth * 100}%;mso-text-raise:${textRaise}" hidden>${zeroWidthSpace.repeat(plSpaceCount)}</i><![endif]-->`}
	</span>
	<span
		style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: {textRaise}px;"
	>
		{@render children()}
	</span>
	<span>
		{@html `<!--[if mso]><i style="mso-font-width:${prFontWidth * 100}%" hidden>${zeroWidthSpace.repeat(prSpaceCount)}&#8203;</i><![endif]-->`}
	</span>
</a>
