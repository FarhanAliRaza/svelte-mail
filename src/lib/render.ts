import type { Component } from 'svelte';
import { render as svelteRender } from 'svelte/server';
import juice from 'juice';
import { convert, type HtmlToTextOptions } from 'html-to-text';
import postcss from 'postcss';
import tailwindcss from 'tailwindcss';
// Try to import @tailwindcss/postcss if available (for Tailwind v4+)
let tailwindPostcss: any;
try {
	// Using dynamic import to handle optional dependency
	tailwindPostcss = require('@tailwindcss/postcss');
} catch (e) {
	// Will use default tailwindcss if @tailwindcss/postcss is not available
}
import type { Config as TailwindConfig } from 'tailwindcss';
import path from 'path';
import fs from 'fs';
// import { marked } from 'marked'; // Import if/when needed

/**
 * Represents the options for the render function.
 */
export interface RenderOptions {
	pretty?: boolean; // Format output HTML
	plainText?: boolean; // Generate plain text version
	tailwindConfig?: string | TailwindConfig; // Path or object for Tailwind configuration
	globalStyles?: string; // Add option for global styles (e.g., fonts)
	htmlToTextOptions?: HtmlToTextOptions; // Options for html-to-text
	juiceOptions?: juice.Options; // Options for juice
}

/**
 * Process HTML with Tailwind CSS.
 * Extracts classes, runs them through Tailwind JIT, and returns the generated CSS.
 *
 * @param html The HTML to process
 * @param tailwindConfig Path to tailwind config or config object
 * @returns The generated CSS
 */
async function processTailwind(
	html: string,
	tailwindConfig?: string | TailwindConfig
): Promise<string> {
	// 1. Extract classes from HTML
	const classRegex = /class="([^"]*)"/g;
	const matches = html.matchAll(classRegex);
	const classSet = new Set<string>();

	for (const match of matches) {
		const classes = match[1].split(/\s+/).filter(Boolean);
		classes.forEach((cls) => classSet.add(cls));
	}

	if (classSet.size === 0) {
		return ''; // No Tailwind classes found
	}

	// 2. Create a temporary CSS file that uses the extracted classes
	const safelist = [...classSet];
	const tempCss = safelist.map((cls) => `.${cls} {}`).join('\n');

	// 3. Load the Tailwind config
	let resolvedConfig: Partial<TailwindConfig> = {};

	if (typeof tailwindConfig === 'string') {
		const configPath = path.resolve(process.cwd(), tailwindConfig);

		if (fs.existsSync(configPath)) {
			try {
				// eslint-disable-next-line @typescript-eslint/no-var-requires
				resolvedConfig = require(configPath);
			} catch (error) {
				console.warn(`Failed to load Tailwind config from ${configPath}:`, error);
			}
		} else {
			console.warn(`Tailwind config file not found at ${configPath}`);
		}
	} else if (tailwindConfig) {
		resolvedConfig = tailwindConfig;
	}

	// 4. Process with Tailwind
	try {
		let postcssPlugins;

		// Use @tailwindcss/postcss if available (for Tailwind v4+), otherwise use direct tailwindcss
		if (tailwindPostcss) {
			postcssPlugins = [
				tailwindPostcss({
					content: [{ raw: html, extension: 'html' }],
					...resolvedConfig
				})
			];
		} else {
			postcssPlugins = [
				tailwindcss({
					content: [{ raw: html, extension: 'html' }],
					...resolvedConfig
				} as any)
			];
		}

		// Process the CSS through postcss
		const result = await postcss(postcssPlugins).process(tempCss, { from: undefined });

		return result.css;
	} catch (error) {
		console.error('Failed to process Tailwind CSS:', error);
		return '';
	}
}

/**
 * Generate plain text version from HTML.
 *
 * @param html The HTML to convert
 * @param options Options for html-to-text
 * @returns The plain text version
 */
function generatePlainText(html: string, options?: HtmlToTextOptions): string {
	const defaultOptions: HtmlToTextOptions = {
		selectors: [
			{ selector: 'head', format: 'skip' },
			{ selector: 'img', format: 'skip' },
			{ selector: 'a', options: { hideLinkHrefIfSameAsText: true } }
		],
		wordwrap: 80
	};

	return convert(html, { ...defaultOptions, ...options });
}

// Remove Svelte/React SSR comments from HTML
function removeSsrComments(html: string): string {
	return html
		.replace(/<!---->/g, '')
		.replace(/<!--\[-->/g, '')
		.replace(/<!--\]-->/g, '');
}

// Remove Svelte/React SSR comments from HTML and other artifacts
function cleanupHtml(html: string): string {
	return html
		.replace(/<!---->/g, '') // Svelte empty comment
		.replace(/<!--\[-->/g, '') // Svelte block start comment
		.replace(/<!--\]-->/g, '') // Svelte block end comment
		.replace(/<!--\[!-->/g, '') // Other potential comment artifacts
		.replace(/\s+onload="this\.__e=event"/g, '') // Remove onload handlers
		.replace(/\s+onerror="this\.__e=event"/g, '') // Remove onerror handlers
		.replace(/\s+class(?=\s|>)/g, '') // Remove empty class attributes (like class or class="")
		.replace(/;\s*undefined/g, ''); // Remove trailing undefined in styles
}

/**
 * Renders a Svelte email component to an HTML string.
 *
 * @param component The root Svelte component to render (e.g., Html).
 * @param options Optional rendering configuration.
 * @param props Props to pass to the component
 * @returns The rendered HTML string or an object with html and text properties if plainText is true.
 */
export async function render<Props extends Record<string, any>>(
	component: Component<any>,
	options?: RenderOptions,
	props?: Props
): Promise<string | { html: string; text: string }> {
	// 1. Render the component using Svelte 5 SSR
	// Svelte 5 returns { head, body }
	const { head, body } = svelteRender(component as any, { props });

	// 2. Combine head and body - Assume the component provides <html> tag
	let combinedHtml = `<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">\n${head}\n${body}`;

	// 3. Process Tailwind CSS if option is provided
	let tailwindGeneratedCss = '';
	if (options?.tailwindConfig) {
		try {
			tailwindGeneratedCss = await processTailwind(combinedHtml, options.tailwindConfig);
		} catch (error) {
			console.warn('Failed to process Tailwind CSS:', error);
		}
	}

	// Combine Tailwind CSS with other global styles (like fonts)
	const allExtraCss = [options?.globalStyles, tailwindGeneratedCss].filter(Boolean).join('\n');

	// 4. Inline CSS using juice
	const juiceOptions: juice.Options = {
		extraCss: allExtraCss, // Use combined global and Tailwind CSS
		removeStyleTags: true, // Remove other style tags potentially added by Svelte
		...options?.juiceOptions
	};

	const inlinedHtml = juice(combinedHtml, juiceOptions);

	// 5. Apply pretty printing if requested
	let finalHtml = inlinedHtml;
	if (options?.pretty) {
		try {
			// Basic pretty printing (consider a more robust library like prettier if needed)
			finalHtml = inlinedHtml
				.replace(/</g, '\n<')
				.replace(/>/g, '>\n')
				.split('\n')
				.map((line) => line.trim())
				.filter(Boolean)
				.join('\n');
		} catch (error) {
			console.warn('Pretty printing failed:', error);
		}
	}

	// Remove Svelte/React SSR comments
	finalHtml = removeSsrComments(finalHtml);

	// Cleanup HTML artifacts
	finalHtml = cleanupHtml(finalHtml);

	// 6. Generate plain text version if requested
	if (options?.plainText) {
		const plainText = generatePlainText(finalHtml, options.htmlToTextOptions);
		return { html: finalHtml, text: plainText };
	}

	return finalHtml;
}

