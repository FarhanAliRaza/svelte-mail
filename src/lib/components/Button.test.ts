// @vitest-environment node
import { describe, it, expect } from 'vitest';
import Button from './Button.svelte';
import type { Snippet } from 'svelte';
import { render } from '../render';

// Create a mock snippet function
const createSnippet = (text: string): Snippet => {
	return (() => text) as unknown as Snippet;
};

describe('Button component (SSR)', () => {
	it('renders with default props', async () => {
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			children: createSnippet('Click me')
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('href="https://example.com"');
		expect(html).toContain('target="_blank"');
		expect(html).toContain('rel="noopener noreferrer"');
		expect(html).toMatchSnapshot();
	});

	it('applies custom styles and props', async () => {
		const customStyle = 'background-color: #ff0000; color: #ffffff;';
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			children: createSnippet('Click me'),
			backgroundColor: '#ff0000',
			color: '#ffffff',
			borderRadius: '8px',
			width: '200px',
			style: customStyle
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('background-color: #ff0000');
		expect(html).toContain('color: #ffffff');
		expect(html).toContain('width="200px"');
		expect(html).toMatchSnapshot();
	});

	it('forwards additional HTML attributes', async () => {
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			children: createSnippet('Click me'),
			'aria-label': 'Click this button',
			'data-testid': 'test-button',
			title: 'Button Title'
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('aria-label="Click this button"');
		expect(html).toContain('data-testid="test-button"');
		expect(html).toContain('title="Button Title"');
		expect(html).toMatchSnapshot();
	});

	it('renders correctly with padding values from padding prop', async () => {
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			padding: '12px 20px',
			children: createSnippet('Click me')
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('padding: 12px 20px 12px 20px');
		expect(html).toContain('[if mso]');
		expect(html).toContain('mso-text-raise:18');
		expect(html).toMatchSnapshot();
	});

	it('renders correctly with explicit padding values', async () => {
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			paddingTop: '10px',
			paddingRight: '15px',
			paddingBottom: '10px',
			paddingLeft: '15px',
			children: createSnippet('Click me')
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('padding: 10px 15px 10px 15px');
		expect(html).toContain('mso-text-raise:15');
		expect(html).toMatchSnapshot();
	});

	it('renders with no padding when not specified', async () => {
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			children: createSnippet('Click me')
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('padding: 0px 0px 0px 0px');
		expect(html).toContain('mso-font-width:0%');
		expect(html).toMatchSnapshot();
	});

	it('allows overriding default style properties', async () => {
		const overridingStyle = `
      line-height: 150%;
      display: block;
      text-decoration: underline;
      max-width: 50%;
    `;
		const result = await render(Button, undefined, {
			href: 'https://example.com',
			style: overridingStyle,
			children: createSnippet('Click me')
		});
		const html = typeof result === 'string' ? result : result.html;
		expect(html).toContain('line-height: 150%');
		expect(html).toContain('display: block');
		expect(html).toContain('text-decoration: underline');
		expect(html).toContain('max-width: 50%');
		expect(html).toMatchSnapshot();
	});
});
