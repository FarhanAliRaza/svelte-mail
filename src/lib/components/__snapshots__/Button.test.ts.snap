// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component (SSR) > allows overriding default style properties 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 150%; text-decoration: underline; display: block; max-width: 50%; mso-padding-alt: 0px; padding: 0px 0px 0px 0px;"><span><!--10kjb78--><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 0px;"></span> <span><!--56ydbm--><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > applies custom styles and props 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 0px 0px 0px 0px; background-color: #ff0000; color: #ffffff;" backgroundcolor="#ff0000" color="#ffffff" borderradius="8px" width="200px"><span><!--10kjb78--><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 0px;"></span> <span><!--56ydbm--><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > forwards additional HTML attributes 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 0px 0px 0px 0px;" aria-label="Click this button" data-testid="test-button" title="Button Title"><span><!--10kjb78--><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 0px;"></span> <span><!--56ydbm--><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > renders correctly with explicit padding values 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 10px 15px 10px 15px;"><span><!--xyluz5--><!--[if mso]><i style="mso-font-width:375%;mso-text-raise:15" hidden>&#8202;&#8202;</i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 15px;"></span> <span><!--1qo1ivl--><!--[if mso]><i style="mso-font-width:375%" hidden>&#8202;&#8202;&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > renders correctly with padding values from padding prop 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 12px 20px 12px 20px;"><span><!--1x4s7i4--><!--[if mso]><i style="mso-font-width:500%;mso-text-raise:18" hidden>&#8202;&#8202;</i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 18px;"></span> <span><!--11jg0db--><!--[if mso]><i style="mso-font-width:500%" hidden>&#8202;&#8202;&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > renders with default props 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 0px 0px 0px 0px;"><span><!--10kjb78--><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 0px;"></span> <span><!--56ydbm--><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>"
`;

exports[`Button component (SSR) > renders with no padding when not specified 1`] = `
"<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<a href="https://example.com" target="_blank" rel="noopener noreferrer" style="line-height: 100%; text-decoration: none; display: inline-block; max-width: 100%; mso-padding-alt: 0px; padding: 0px 0px 0px 0px;"><span><!--10kjb78--><!--[if mso]><i style="mso-font-width:0%;mso-text-raise:0" hidden></i><![endif]--></span> <span style="max-width: 100%; display: inline-block; line-height: 120%; mso-padding-alt: 0px; mso-text-raise: 0px;"></span> <span><!--56ydbm--><!--[if mso]><i style="mso-font-width:0%" hidden>&#8203;</i><![endif]--></span></a>"
`;
